File: Project/GroupTableData.cpp
#include "GroupTableData.hpp"

const std::string GroupTableData::FIO_COLUMN_HEADER = "ФИО";
const std::string GroupTableData::AVERAGE_COLUMN_HEADER = "Средний балл";
const int GroupTableData::EXTRA_COLUMNS_COUNT = 2;


GroupTableData::GroupTableData(const StudentArray& students, const DisciplineArray& disciplines,
	const AttestationResultTable& tableBody) :
	m_students(students), m_disciplines(disciplines), m_tableBody(tableBody) { }

GroupTableData::GroupTableData(StudentArray&& students, DisciplineArray&& disciplines,
	AttestationResultTable&& tableBody) :
	m_students(std::move(students)), m_disciplines(std::move(disciplines)), m_tableBody(std::move(tableBody)) { }

GroupTableData::GroupTableData(const GroupTableData& other) :
	m_students(other.m_students), m_disciplines(other.m_disciplines), m_tableBody(other.m_tableBody) { }

const GroupTableData::StudentArray& GroupTableData::GetStudents() const {
	return m_students;
}

const GroupTableData::DisciplineArray& GroupTableData::GetDisciplines() const {
	return m_disciplines;
}

const GroupTableData::AttestationResultTable& GroupTableData::GetTableBody() const {
	return m_tableBody;
}

std::vector<int> GroupTableData::StudentAverages() const {
	std::vector<int> studentAverages(m_students.size(), 0);
	for (int i = 0; i < m_students.size(); i++) {
		int k = 0;
		for (int j = 0; j < m_disciplines.size(); j++) {
			if (m_disciplines[j]->GetAttestationType() == Discipline::AttestationType::eExam) {
				studentAverages[i] += m_tableBody[i][j]->ToPercent();
				k += 1;
			}
		}
		if (k != 0) {
			studentAverages[i] /= k;
		}
	}
	return studentAverages;
}

std::vector<int> GroupTableData::DisciplineAverages() const {
	std::vector<int> disciplineAverages(m_disciplines.size(), 0);
	for (int i = 0; i < m_disciplines.size(); i++) {
		for (int j = 0; j < m_students.size(); j++) {
			disciplineAverages[i] += m_tableBody[j][i] ? m_tableBody[j][i]->ToPercent() : 0;
		}
		if (m_students.size() != 0) {
			disciplineAverages[i] /= m_students.size();
		}
	}
	return disciplineAverages;
}

int GroupTableData::GroupAverage() const {
	std::vector<int> studentAverages = StudentAverages();
	int res = 0;
	for (int i = 0; i < m_students.size(); i++) {
		res += studentAverages[i];
	}
	if (m_students.size() == 0) {
		return 0;
	}
	return res / m_students.size();
}


std::vector<std::string> GroupTableData::TableHeader() const {
	std::vector<std::string> cells;
	cells.push_back(FIO_COLUMN_HEADER);
	for (int i = 1; i <= m_disciplines.size(); i++) {
		cells.push_back(std::to_string(i));
	}
	cells.push_back(AVERAGE_COLUMN_HEADER);

	return cells;
}

std::vector<std::vector<std::string>> GroupTableData::TableBodyToStrings(bool compactResults) const {
	std::vector<std::vector<std::string>> result;

	std::vector<int> studentAverages = StudentAverages();
	for (int i = 0; i < m_students.size(); i++) {
		std::vector<std::string> cells(m_disciplines.size() + EXTRA_COLUMNS_COUNT);

		cells[0] = m_students[i]->GetLastNameWithInitials();
		for (int j = 0; j < m_disciplines.size(); j++) {
			const std::shared_ptr<AttestationResult>& res = m_tableBody[i][j];
			if (compactResults) {
				cells[j + 1] = res ? res->ToStringCompact() : "";
			} else {
				cells[j + 1] = res ? res->ToString() : "";
			}
		}
		cells.back() = std::to_string(studentAverages[i]);

		result.push_back(cells);
	}

	return result;
}

std::vector<std::string> GroupTableData::TableDisciplineAverages() const {
	std::vector<std::string> cells;
	cells.push_back("");
	for (int average : DisciplineAverages()) {
		cells.push_back(std::to_string(average));
	}
	cells.push_back(std::to_string(GroupAverage()));

	return cells;
}

File: ProjectCS/GroupTableData.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SessionResultsCS
{
    public class GroupTableData
    {
        public const string FIO_COLUMN_HEADER = "ФИО";
        public const string AVERAGE_COLUMN_HEADER = "Средний балл";
        public const int EXTRA_COLUMNS_COUNT = 2;


        // Массив студентов в порядке вхождения в таблицу
        public Student[] Students { get; }

        // Массив дисциплин в порядке вхождения в таблицу
        public Discipline[] Disciplines { get; }

        // Двумерный массив результатов аттестации
        public AttestationResult?[,] TableBody { get; }


        public GroupTableData(Student[] students, Discipline[] disciplines,
            AttestationResult?[,] tableBody)
        {
            if (tableBody.GetLength(0) != students.Length || tableBody.GetLength(1) != disciplines.Length)
            {
                throw new ArgumentException("Invalid tableData size");
            }
            Students = students;
            Disciplines = disciplines;
            TableBody = tableBody;
            
        }


        // Возвращает массив средних баллов студентов по экзаменам
        public int[] StudentAverages()
        {
            int[] studentAverages = new int[Students.Length];
            for (int i = 0; i < Students.Length; i++)
            {
                int k = 0;
                for (int j = 0; j < Disciplines.Length; j++)
                {
                    if (Disciplines[j].AttestationType == AttestationType.Exam)
                    {
                        studentAverages[i] += TableBody[i, j]?.ToScore() ?? 0;
                        k += 1;
                    }
                }
                if (k != 0)
                {
                    studentAverages[i] /= k;
                }
            }

            return studentAverages;
        }

        // Возвращает массив средних баллов/процентов сдачи по дисциплинам
        public int[] DisciplineAverages()
        {
            int[] disciplineAverages = new int[Disciplines.Length];
            for (int j = 0; j < Disciplines.Length; j++)
            {
                for (int i = 0; i < Students.Length; i++)
                {
                    disciplineAverages[j] += TableBody[i, j]?.ToScore() ?? 0;
                }
                if (Students.Length != 0)
                {
                    disciplineAverages[j] /= Students.Length;
                }
            }

            return disciplineAverages;
        }


        public int GroupAverage()
        {
            return (int)StudentAverages().Average();
        }


        public string[] TableHeader()
        {
            string[] cells = new string[Disciplines.Length + EXTRA_COLUMNS_COUNT];
            cells[0] = FIO_COLUMN_HEADER;
            for (int i = 1; i <= Disciplines.Length; i++)
            {
                cells[i] = i.ToString();
            }
            cells[cells.Length - 1] = AVERAGE_COLUMN_HEADER;

            return cells;
        }

        public string[][] TableBodyToStrings(bool compactResults)
        {
            string[][] result = new string[Students.Length][];

            int[] studentAverages = StudentAverages();
            for (int i = 0; i < Students.Length; i++)
            {
                string[] cells = new string[Disciplines.Length + EXTRA_COLUMNS_COUNT];

                cells[0] = Students[i].GetLastNameWithInitials();
                for (int j = 0; j < Disciplines.Length; j++)
                {
                    if (compactResults)
                    {
                        cells[j + 1] = TableBody[i, j]?.ToStringCompact() ?? "";
                    }
                    else
                    {
                        cells[j + 1] = TableBody[i, j]?.ToString() ?? "";
                    }
                }
                cells[cells.Length - 1] = studentAverages[i].ToString();

                result[i] = cells;
            }

            return result;
        }

        public string[] TableDisciplineAverages()
        {
            string[] cells = new string[Disciplines.Length + EXTRA_COLUMNS_COUNT];

            cells[0] = "";
            int[] disciplineAverages = DisciplineAverages();
            for (int i = 1; i <= Disciplines.Length; i++)
            {
                cells[i] = disciplineAverages[i - 1].ToString();
            }
            cells[cells.Length - 1] = GroupAverage().ToString();

            return cells;
        }
    }
}
